localstack/services/s3/utils.py

```diff
@@ -0,0 +1,46 @@
+from typing import Type, TypedDict
+
+from localstack.aws.api import ServiceException
+from localstack.aws.api.s3 import ChecksumAlgorithm, PutObjectRequest
+from localstack.utils.strings import checksum_crc32, checksum_crc32c, hash_sha1, hash_sha256
+
+checksum_keys = ["ChecksumSHA1", "ChecksumSHA256", "ChecksumCRC32", "ChecksumCRC32C"]
+
+
+class InvalidRequest(ServiceException):
+    """The lifecycle configuration does not exist."""
+
+    code: str = "InvalidRequest"
+    sender_fault: bool = False
+    status_code: int = 400
+
+
+def get_all_typed_dict_keys(typed_dict: Type[TypedDict]):
+    return [*typed_dict.__required_keys__, *typed_dict.__optional_keys__]
+
+
+def verify_checksum(checksum_algorithm: str, data: bytes, request: PutObjectRequest):
```

Comment:
I think this should be a `Handler` implementation in the AWS handler chain (but this could also be done in the next iteration).
The spec for `PutObject` looks like this:
```
    "PutObject":{
      "name":"PutObject",
      ...
      "httpChecksum":{
        "requestAlgorithmMember":"ChecksumAlgorithm",
        "requestChecksumRequired":false
      }
    },
```

The `httpChecksum` field can only be found 38 times, but only in the specs for `s3` and `s3control`.
I guess it would be fairly simple to have a handler which checks if `context.operation.http_checksum` is set (and if so enforces it).
Here's a documentation on the Smithy trait: https://awslabs.github.io/smithy/2.0/aws/aws-core.html#aws-protocols-httpchecksum-trait
What's your take on this, @thrau?

Reply:
Right, this makes sense! We can introduce this in the next iteration, that would be great to have it abstracted away!